### ISSUE TITLE
Clustering with the CNTK?

### ISSUE THREAD

**Issue by PascalS86** [Microsoft/CNTK#2743]
Hi everybody? are there plans for clustering algorithm within the cntk? An implementation of (e.g.) k-means like in Accord-Framework or in the azure methods in Cognitive Services, would be awesome.

**Comment by rhy-ama** [comment]
The best would be implement a sample/tutorial for SOM ([wiki:Self-Organising Maps](https://en.wikipedia.org/wiki/Self-organizing_map)) - which is a generalization of clustering algo family.

Maybe a link to the reproducible implementation (academic research or otherwise) would help the team to include this into features. 
- [tensorflow implementation](https://codesachin.wordpress.com/2015/11/28/self-organizing-maps-with-googles-tensorflow/)
- one interesting paper: [Principal temporal extensions of SOM](http://www.sersc.org/journals/IJSIP/vol2_no3/10.pdf)

**Comment by ArchiDevil** [comment]
You can use something like autoencoder with any clustering algorithm from scikit. It is already on your machine if you installed Anaconda.

**Comment by PascalS86** [comment]
Hey, thanks for your response. It's not about orchestrating different frameworks. I know, in python I can use scikit or tensorflow to achieve this. And in C# I can use Accord.NET to do this. I just think, that this is missing in cntk. So there are no plans to add this feature to CNTK?

**Comment by n17s** [comment]
The kmeans objective can be written down as a network and you can then use your favorite learner:
```python
x = C.input_variable((1,dimension)
c = C.parameter((dimension, num_clusters))
loss = C.reduce_min(C.reduce_sum(c*c, axis=0) - 2 * C.times(x, c))
```
which basically says that the loss for a point and a bunch of centroids is the minimum of distances from all centroids. For the distance I have used ||x-c||^2 = ||c||^2 - 2*dot(x,c) + constant.
